File: openc3/python/openc3/interfaces/protocols/slip_protocol.py
# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

from openc3.config.config_parser import ConfigParser
from openc3.interfaces.protocols.terminated_protocol import TerminatedProtocol

# This file implements the SLIP protocol as documented in RFC 1055
# https://datatracker.ietf.org/doc/html/rfc1055

# SLIP is a framing protocol and is therefore expected to be used for packet deliniation


class SlipProtocol(TerminatedProtocol):
    # Note: Characters are expected to be given as integers
    # @param start_char   [Integer/nil] Character to place at the start of frames (Defaults to nil)
    # @param read_strip_characters [true/false] Strip off start_char and end_char from reads
    # @param read_enable_escaping [true/false] Whether to enable or disable character escaping on reads
    # @param write_enable_escaping [true/false] Whether to enable or disable character escaping on writes
    # @param end_char     [Integer] Character to place at the end of frames (Defaults to 0xC0)
    # @param esc_char     [Integer] Escape character (Defaults to 0xDB)
    # @param esc_end_char [Integer] Character to Escape End character (Defaults to 0xDC)
    # @param esc_esc_char [Integer] Character to Escape Escape character (Defaults to 0xDD)
    # @param allow_empty_data [true/false/nil] See Protocol#initialize
    def __init__(
        self,
        start_char=None,
        read_strip_characters=True,
        read_enable_escaping=True,
        write_enable_escaping=True,
        end_char="0xC0",
        esc_char="0xDB",
        esc_end_char="0xDC",
        esc_esc_char="0xDD",
        allow_empty_data=None,
    ):
        self.start_char = ConfigParser.handle_none(start_char)
        if self.start_char is not None:
            try:
                self.start_char = int(start_char, 0).to_bytes(1, byteorder="big")
            except ValueError:
                raise ValueError(f"invalid value {start_char} for start_char")
        self.end_char = int(end_char, 0).to_bytes(1, byteorder="big")
        self.esc_char = int(esc_char, 0).to_bytes(1, byteorder="big")
        self.esc_esc_char = int(esc_esc_char, 0).to_bytes(1, byteorder="big")
        self.esc_end_char = int(esc_end_char, 0).to_bytes(1, byteorder="big")
        self.replace_end = self.esc_char + self.esc_end_char
        self.replace_esc = self.esc_char + self.esc_esc_char
        self.read_strip_characters = ConfigParser.handle_true_false(
            read_strip_characters
        )
        if (
            self.read_strip_characters is not True
            and self.read_strip_characters is not False
        ):
            raise RuntimeError("read_strip_characters must be True or False")
        self.read_enable_escaping = ConfigParser.handle_true_false(read_enable_escaping)
        if (
            self.read_enable_escaping is not True
            and self.read_enable_escaping is not False
        ):
            raise RuntimeError("read_enable_escaping must be True or False")
        self.write_enable_escaping = ConfigParser.handle_true_false(
            write_enable_escaping
        )
        if (
            self.write_enable_escaping is not True
            and self.write_enable_escaping is not False
        ):
            raise RuntimeError("write_enable_escaping must be True or False")

        strip_read_termination = False
        discard_leading_bytes = 0
        sync_pattern = None
        if self.start_char:
            sync_pattern = hex(int(start_char, 0))
        fill_fields = False  # Handled in write_data below

        super().__init__(
            "",  # Write termination handled in write_data below
            hex(int(end_char, 0)),  # Expects Hex Character String
            strip_read_termination,
            discard_leading_bytes,
            sync_pattern,
            fill_fields,
            allow_empty_data,
        )

    def read_data(self, data, extra=None):
        data, extra = super().read_data(data, extra)
        if len(data) <= 0 or type(data) is str:
            return (data, extra)

        if self.read_strip_characters:
            if self.start_char:
                data = data[1:]
            data = data[0:-1]

        if self.read_enable_escaping:
            data = data.replace(self.replace_end, self.end_char).replace(
                self.replace_esc, self.esc_char
            )

        return (data, extra)

    def write_data(self, data, extra=None):
        # Intentionally not calling super()

        if self.write_enable_escaping:
            data = data.replace(self.esc_char, self.replace_esc).replace(
                self.end_char, self.replace_end
            )

        if self.start_char:
            data = self.start_char + data

        data += self.end_char

        return (data, extra)

    def reduce_to_single_packet(self, extra=None):
        if len(self.data) <= 0:
            return ("STOP", extra)
        index = None
        if self.start_char is not None:
            try:
                index = self.data[1:].index(self.read_termination_characters)
                index = index + 1
            except ValueError:
                pass
        else:
            try:
                index = self.data.index(self.read_termination_characters)
            except ValueError:
                pass

        # Reduce to packet data and setup current_data for next packet
        if index is not None:
            if index > 0:
                packet_data = self.data[
                    0 : (index + len(self.read_termination_characters))
                ]
            else:  # self.data begins with the termination characters
                packet_data = self.data[0 : (len(self.read_termination_characters))]
            self.data = self.data[(index + len(self.read_termination_characters)) :]
            return (packet_data, extra)
        else:
            return ("STOP", extra)

File: openc3/lib/openc3/interfaces/protocols/slip_protocol.rb
# encoding: ascii-8bit

# Copyright 2023 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/interfaces/protocols/terminated_protocol'
require 'openc3/config/config_parser'

# This file implements the SLIP protocol as documented in RFC 1055
# https://datatracker.ietf.org/doc/html/rfc1055

# SLIP is a framing protocol and is therefore expected to be used for packet deliniation

module OpenC3

  # Usage in plugin.txt:
  #
  # INTERFACE ...
  #   PROTOCOL READ_WRITE SlipProtocol

  class SlipProtocol < TerminatedProtocol

    # Note: Characters are expected to be given as integers
    # @param start_char   [Integer/nil] Character to place at the start of frames (Defaults to nil)
    # @param read_strip_characters [true/false] Strip off start_char and end_char from reads
    # @param read_enable_escaping [true/false] Whether to enable or disable character escaping on reads
    # @param write_enable_escaping [true/false] Whether to enable or disable character escaping on writes
    # @param end_char     [Integer] Character to place at the end of frames (Defaults to 0xC0)
    # @param esc_char     [Integer] Escape character (Defaults to 0xDB)
    # @param esc_end_char [Integer] Character to Escape End character (Defaults to 0xDC)
    # @param esc_esc_char [Integer] Character to Escape Escape character (Defaults to 0xDD)
    # @param allow_empty_data [true/false/nil] See Protocol#initialize
    def initialize(
      start_char = nil,
      read_strip_characters = true,
      read_enable_escaping = true,
      write_enable_escaping = true,
      end_char = 0xC0,
      esc_char = 0xDB,
      esc_end_char = 0xDC,
      esc_esc_char = 0xDD,
      allow_empty_data = nil)

      @start_char = ConfigParser.handle_nil(start_char)
      @start_char = [Integer(start_char)].pack('C') if @start_char
      @end_char = [Integer(end_char)].pack('C')
      @esc_char = [Integer(esc_char)].pack('C')
      @esc_end_char = [Integer(esc_end_char)].pack('C')
      @esc_esc_char = [Integer(esc_esc_char)].pack('C')
      @replace_end = @esc_char + @esc_end_char
      @replace_esc = @esc_char + @esc_esc_char
      @read_strip_characters = ConfigParser.handle_true_false(read_strip_characters)
      raise "read_strip_characters must be true or false" if @read_strip_characters != true and @read_strip_characters != false
      @read_enable_escaping = ConfigParser.handle_true_false(read_enable_escaping)
      raise "read_enable_escaping must be true or false" if @read_enable_escaping != true and @read_enable_escaping != false
      @write_enable_escaping = ConfigParser.handle_true_false(write_enable_escaping)
      raise "write_enable_escaping must be true or false" if @write_enable_escaping != true and @write_enable_escaping != false

      strip_read_termination = false
      discard_leading_bytes = 0
      if @start_char
        sync_pattern = sprintf("%0X", Integer(start_char))
      else
        sync_pattern = nil
      end
      fill_fields = false # Handled in write_data below

      super(
        "", # Write termination handled in write_data below
        sprintf("%0X", Integer(end_char)), # Expects Hex Character String
        strip_read_termination,
        discard_leading_bytes,
        sync_pattern,
        fill_fields,
        allow_empty_data
      )
    end

    def read_data(data, extra = nil)
      data, extra = super(data, extra)
      return data, extra if data.length <= 0 or Symbol === data

      if @read_strip_characters
        if @start_char
          data = data[1..-1]
        end
        data = data[0..-2]
      end

      if @read_enable_escaping
        data = data.gsub(@replace_end, @end_char).gsub(@replace_esc, @esc_char)
      end

      return data, extra
    end

    def write_data(data, extra = nil)
      # Intentionally not calling super()

      if @write_enable_escaping
        data = data.gsub(@esc_char, @replace_esc).gsub(@end_char, @replace_end)
      end

      if @start_char
        data = @start_char + data
      end

      data << @end_char

      return data, extra
    end

    def reduce_to_single_packet
      return :STOP if @data.length <= 0
      if @start_char
        index = @data[1..-1].index(@read_termination_characters)
        index = index + 1 if index
      else
        index = @data.index(@read_termination_characters)
      end

      # Reduce to packet data and setup current_data for next packet
      if index
        if index > 0
          packet_data = @data[0..(index + @read_termination_characters.length - 1)]
        else # @data begins with the termination characters
          packet_data = @data[0..(@read_termination_characters.length - 1)]
        end
        @data.replace(@data[(index + @read_termination_characters.length)..-1])
        return packet_data, @extra
      else
        return :STOP
      end
    end

    def write_details
      result = super()
      result['start_char'] = @start_char.inspect
      result['end_char'] = @end_char.inspect
      result['esc_char'] = @esc_char.inspect
      result['esc_end_char'] = @esc_end_char.inspect
      result['esc_esc_char'] = @esc_esc_char.inspect
      result['write_enable_escaping'] = @write_enable_escaping
      return result
    end

    def read_details
      result = super()
      result['start_char'] = @start_char.inspect
      result['end_char'] = @end_char.inspect
      result['esc_char'] = @esc_char.inspect
      result['esc_end_char'] = @esc_end_char.inspect
      result['esc_esc_char'] = @esc_esc_char.inspect
      result['read_strip_characters'] = @read_strip_characters
      result['read_enable_escaping'] = @read_enable_escaping
      return result
    end
  end

end
